File: kotlin/core/src/main/kotlin/Clazz.kt
package ir.kotlin.core

// define a class named `Person` with a single property `name` of type `String`
class Person(val name: String, var age: Int) {
    // Custom accessors
    val isAdult: Boolean
        get() = age >= 18

    fun greet() = println("Hello, my name is $name and I am $age years old. Is adult: $isAdult")
    
    // needs toString method to print the object in a readable format
    override fun toString(): String {
        return "Person(name='$name', age=$age, isAdult=$isAdult)"
    }
}

fun main() {
    // create an instance of `Person` with a name and age without using `new`
    val person = Person("Alice", 16)
    person.greet()
    
    person.age = 31
    person.greet()
    println("Is adult: ${person}")
}

File: javascript/core/src/clazz.js
// This code demonstrates the use of classes in JavaScript, including inheritance and polymorphism.
// ES6 introduced classes as syntactical sugar over JavaScript's existing prototype-based inheritance.
class Person {
  constructor(name, age) {
    this.name = name;
    this.age = age;
  }

  greet() {
    return `Hello, my name is ${this.name} and I am ${this.age} years old.`;
  }
}

// inheritance
class Student extends Person {
  constructor(name, age, grade) {
    super(name, age);
    this.grade = grade;
  }

  study() {
    return `${this.name} is studying for grade ${this.grade}.`;
  }
}

class Teacher extends Person {
  constructor(name, age, subject) {
    super(name, age);
    this.subject = subject;
  }

  teach() {
    return `${this.name} is teaching ${this.subject}.`;
  }
}

const person = new Person("Alice", 30);
const student = new Student("Bob", 20, "10th");
const teacher = new Teacher("Charlie", 40, "Mathematics");

// polymorphism
console.log(person.greet());
console.log(student.greet());
console.log(teacher.greet());

// javascript do not support overloading, but we can achieve similar behavior using default parameters
function greet(person, greeting = "Hello") {
  return `${greeting}, my name is ${person.name}.`;
}
